libyul/optimiser/Suite.cpp

```diff
@@ -177,7 +184,10 @@ void OptimiserSuite::run(
 
 		{
 			// SSA plus simplify
+			ConditionalSimplifier{_dialect}(ast);
```

Comment:
I think the best sequence for calling the conditional simplifier is:
- ``ExpressionSplitter``
- ``CommonSubexpressionEliminator``
- ``ConditionalSimplifier``
- ``Rematerializer``
- ``RedundantAssignEliminator``

I think this entire sequence together would even retain SSA form, if the rematerializer was restricted to only rematerialize literals.

Reply:
As a rough example (I haven't actually run it, but this is the idea at least):
```
                if iszero(_3) { break }
                if iszero(_3) { revert(_1, _1) }
```
=> ExpressionSplitter
```
                let _4 := iszero(_3)
                if _4 { break }
                let _5 := iszero(_3)
                if _5 { revert(_1, _1) }
```
=> CSE
```
                let _4 := iszero(_3)
                if _4 { break }
                let _5 := _4
                if _5 { revert(_1, _1) }
```
=> ConditionalSimplifier
```
                let _4 := iszero(_3)
                if _4 { break }
                _4 := 0
                let _5 := _4
                if _5 { revert(_1, _1) }
```
=> Rematerializer
```
                let _4 := iszero(_3)
                if _4 { break }
                _4 := 0
                let _5 := 0
                if _5 { revert(_1, _1) }
```
=> RedundantAssignEliminator
```
                let _4 := iszero(_3)
                if _4 { break }
                let _5 := 0
                if _5 { revert(_1, _1) }
```
And the knowledge stemming from the if condition is used and we are even back in SSA form.

Reply:
Hm... in case that ``_4`` is part of the loop condition, this, however, will not restore SSA form... in that case it wasn't SSA form anyways, though, but it might be problematic in cases like the never executed loop with dead code below, but I haven't thought it through yet...

Reply:
@chriseth Isn't this sequence similar to the one we ran manually?